### ISSUE TITLE
Adding os_patching class result in failed runs on all Windows nodes on first run

### ISSUE THREAD

**Issue by genebean** [albatrossflavour/puppet_os_patching#172]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: PE 2019.3.0
- Distribution: Windows servers from 2012 - 2016
- Module version: 0.13.0

## What are you seeing

On first run I had a timeout on all my Windows nodes. Second run worked. All that was done was adding the `os_patching` class.

```
Command exceeded timeout
Source: /Stage[main]/Os_patching/Exec[os_patching::exec::fact]
File: /etc/puppetlabs/code/environments/os_patching/modules/os_patching/manifests/init.pp
Line: 376
```

## What behaviour did you expect instead

I expect the run to work on the first try

## Any additional information you'd like to impart

From talking with others here it sounds like a timeout parameter could be added to the exec at https://github.com/albatrossflavour/puppet_os_patching/blob/e4de14bf70f470ed80f1aaa8c2c6ebb1e2f3c9b7/manifests/init.pp#L376-L380 (a la https://puppet.com/docs/puppet/latest/types/exec.html#exec-attribute-timeout)

On a possibly related note, an engineer I was talking to asked why this script resubmitted facts mid run seeing as they are submitted pre and post run already. Not sure if that is relevant to the issue but figured it was worth mentioning.

**Comment by nathangiuliani** [comment]
You've raised an interesting point there. Part of updating the fact is performing an update scan, which can take a very long time depending on your OS version, which updates are installed, and the configuration of your update source (usually WSUS, but could be Windows/Microsoft update).

Adding a timeout would likely involve forcefully ending the update scan API call which would just cause this to happen continually each puppet run, until a scan was completed outside of this process (e.g. the background scans that the windows update agent can perform).

The best 'fix' is to make the update scan complete quickly, but this can be a lot of work on the WSUS server, or sometimes impossible (e.g. Windows 8 or Server 2012 have a bug that causes scans to take a long time that was really only fixed in 8.1/2012R2).

We could also look at using the asynchronous scan APIs or a scheduled task (far easier to implement but less reliable) to update the fact data instead of performing the scan as part of an update run. The code used currently is currently a "like for like" for the linux model that this module started with where yum or apt was called to do a 'scan' each time the fact was updated.

**Comment by genebean** [comment]
What about if, as a first pass, a timeout setting is added so that it can run longer than the default 300 seconds? My issue was that Puppet's exec resource timed out. Alternatively, maybe a timeout parameter is added to the resource defined at https://github.com/albatrossflavour/puppet_os_patching/blob/e4de14bf70f470ed80f1aaa8c2c6ebb1e2f3c9b7/manifests/init.pp#L376-L380 so that it becomes 

```puppet
 exec { $fact_exec: 
   path        => 'C:/Windows/System32/WindowsPowerShell/v1.0', 
   timeout     => 900,
   refreshonly => true, 
   command     => "powershell -executionpolicy remotesigned -file ${fact_cmd}", 
 } 
```

**Comment by nathangiuliani** [comment]
It's been a while since I looked into this, but from memory I couldn't find a way to implement a timeout on the API call in the PowerShell code. Increasing the timeout of the exec resource is probably the best option at this stage?

**Comment by genebean** [comment]
I think it at least leads to a better user experience. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by albatrossflavour** [comment]
Timeout added into development and will be released in 0.16.0
